### PR TITLE
Fix crash when setting default on form field properties

### LINKED ISSUE #57121: QGIS crashes when using default value "now()"
### What is the bug or the crash?

In QGIS 3.36(.1) QGIS crashes when using a "now()"-default value.

Bug started in 3.36.x

### Steps to reproduce the issue

1)
Create a new user-profile
2) 
Create a new gpkg-layer with some text field and date-fields
3) 
go to layer-properties-> set a default value "now()" to one of the fields
4) 
select another field
5) 
crash
grafik

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.36.1-Maidenhead | QGIS code revision | 3e589453
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.18
GDAL/OGR version | 3.8.4
PROJ version | 9.3.1
EPSG Registry database version | v10.098 (2023-11-24)
GEOS version | 3.12.1-CAPI-1.18.1
SQLite version | 3.41.1
PDAL version | 2.6.0
PostgreSQL client version | 16.2
SpatiaLite version | 5.1.0
QWT version | 6.1.6
QScintilla2 version | 2.13.4
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99

</body></html>QGIS version
3.36.1-Maidenhead
QGIS code revision
[3e589453](https://github.com/qgis/QGIS/commit/3e589453)
Qt version
5.15.3
Python version
3.9.18
GDAL/OGR version
3.8.4
PROJ version
9.3.1
EPSG Registry database version
v10.098 (2023-11-24)
GEOS version
3.12.1-CAPI-1.18.1
SQLite version
3.41.1
PDAL version
2.6.0
PostgreSQL client version
16.2
SpatiaLite version
5.1.0
QWT version
6.1.6
QScintilla2 version
2.13.4
OS version
Windows 10 Version 2009

Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by elpaso**
Fix #57121

Founded by: Take the A train ♫

### CODE REVIEW

File: src/gui/vector/qgsattributesformproperties.cpp [hunk 1]
```diff
@@ -378,16 +378,17 @@ void QgsAttributesFormProperties::storeAttributeTypeDialog()
...
-
-  mLayer->setDefaultValueDefinition( mAttributeTypeDialog->fieldIdx(), QgsDefaultValue( mAttributeTypeDialog->defaultValueExpression(), mAttributeTypeDialog->applyDefaultValueOnUpdate() ) );
-
   cfg.mEditorWidgetType = mAttributeTypeDialog->editorWidgetType();
   cfg.mEditorWidgetConfig = mAttributeTypeDialog->editorWidgetConfig();
-
   cfg.mSplitPolicy = mAttributeTypeDialog->splitPolicy();
+  const int fieldIndex = mAttributeTypeDialog->fieldIdx();
+
+  mLayer->setDefaultValueDefinition( mAttributeTypeDialog->fieldIdx(), QgsDefaultValue( mAttributeTypeDialog->defaultValueExpression(), mAttributeTypeDialog->applyDefaultValueOnUpdate() ) );
```
Comment (m-kuhn) [inline src/gui/vector/qgsattributesformproperties.cpp:389]:
```suggestion
  mLayer->setDefaultValueDefinition( fieldIdx, QgsDefaultValue( mAttributeTypeDialog->defaultValueExpression(), mAttributeTypeDialog->applyDefaultValueOnUpdate() ) );
```

**Review by m-kuhn — APPROVED** [review]
I have no idea why it was crashing before but the code looks good

**Comment by elpaso** [comment]
> I have no idea why it was crashing before but the code looks good

I tried to explain the reason for the crash in the comment here https://github.com/qgis/QGIS/pull/57126/files#diff-8815a88114ddfdb7fe231a22c98e88f4fa4bba1ba58d068ae549413021034dacR381 but I apparently failed because that is just the ultimate cause of the crash and also because it wasn't entirely correct, the dialog is deleted "later" but the member variable is set to `nullptr` immediately: the call stack is:

image


I have the feeling that the root cause of the bug was introduced by https://github.com/qgis/QGIS/commit/9c78dec2b6068b31804d30375b75befa4869c07b and then https://github.com/qgis/QGIS/commit/086d5cfa28ea1427d1854bdc0b9d0bffc9a43111#diff-8815a88114ddfdb7fe231a22c98e88f4fa4bba1ba58d068ae549413021034dacR90 but I'm not sure.

**Comment by m-kuhn** [comment]
Thanks for the explanation. 👍🏿

**Pull request merged**